### PR TITLE
#10834: Overload backward ops

### LINKED ISSUE #10834: Set 1 ops
Tracking Issue : #10833 
List of ops : #10771

- [x] Add 
- [x] LT
- [x] Sub
- [x] RECIP
- [x] EQ
- [x] GT
- [x] ABS

### PR DESCRIPTION
**Pull request by Aswinmcw**
Overload backward ops 
Issue : #10834 
### Checklist
- [ ] Post commit CI passes
- [ ] Model regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/eltwise/binary_backward/device/binary_backward_op.cpp [hunk 1]
```diff
@@ -176,6 +176,26 @@ std::vector<ComplexTensor> ExecuteUnaryBackwardAdd::operator()(
     return grad_tensor;
 }
 
+std::vector<Tensor> ExecuteUnaryBackwardComparison::operator()(
+    const Tensor& grad, const Tensor& input, float alpha, const std::optional<MemoryConfig>& output_mem_config) {
+    cout<<"HEREEE \n";
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/eltwise/binary_backward/device/binary_backward_op.cpp:181]: cout<<"HEREEE \n";

File: ttnn/cpp/ttnn/operations/eltwise/binary_backward/binary_backward_pybind.hpp [hunk 2]
```diff
@@ -402,10 +403,168 @@ Keyword args:
             py::arg("memory_config") = std::nullopt});
 }
 
+template <typename binary_backward_operation_t>
+void bind_binary_bw_operation(py::module& module, const binary_backward_operation_t& operation, const std::string& description) {
+    auto doc = fmt::format(
+        R"doc({0}(input_tensor_a: ttnn.Tensor, input_tensor_b: Union[ttnn.Tensor, int, float], *, memory_config: Optional[ttnn.MemoryConfig] = None, dtype: Optional[ttnn.DataType] = None, activations: Optional[List[str]] = None) -> ttnn.Tensor
```
Comment (VirdhatchaniKN) [inline ttnn/cpp/ttnn/operations/eltwise/binary_backward/binary_backward_pybind.hpp:409]: Also a minor Update here : `input_tensor_a` and  `input_tensor_b` can be of `ComplexTensor `.

File: ttnn/cpp/ttnn/operations/eltwise/unary_backward/device/unary_backward_op.hpp [hunk 3]
```diff
@@ -783,6 +784,9 @@ struct OpHandler<UnaryBackwardOpType::ADD_BW> {
     static std::vector<Tensor> handle( const Tensor& grad, const Tensor& input, float alpha, const std::optional<MemoryConfig>& output_mem_config ) {
         return _add_bw(grad, input, alpha, output_mem_config);
     }
+    static std::vector<Tensor> handle( const Tensor& grad, const Tensor& input_a, const Tensor& input_b, const std::optional<MemoryConfig>& output_mem_config ) {
```
Comment (VirdhatchaniKN) [inline ttnn/cpp/ttnn/operations/eltwise/unary_backward/device/unary_backward_op.hpp:787]: Is `_add_bw_inter` needed?

File: ttnn/cpp/ttnn/operations/eltwise/unary_backward/device/unary_backward_op.cpp [hunk 4]
```diff
@@ -19,8 +19,13 @@
 #include "ttnn/operations/reduction/prod/prod.hpp"
 #include "ttnn/operations/eltwise/ternary/where_op.hpp"
 #include "ttnn/operations/eltwise/unary/unary_composite.hpp"
+#include "ttnn/operations/eltwise/complex/complex.hpp"
+#include "ttnn/operations/eltwise/unary_backward/unary_backward.hpp"
+#include "ttnn/operations/eltwise/complex_unary/complex_unary.hpp"
+#include "ttnn/operations/eltwise/complex_binary_backward/device/complex_binary_backward_op.hpp"
 
 namespace ttnn::operations::unary_backward {
+using ComplexTensor = complex::ComplexTensor;
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/eltwise/unary_backward/device/unary_backward_op.cpp:28]: Maybe just define ComplexTensor in ttnn instead, so you don't have to do this.

**Review by ayerofieiev-tt — APPROVED** [review]
Some more .hpp should be split into .hpp and .cpp, but I am not forcing this in current PR. Overall looks good

File: ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward_pybind.hpp [hunk 5]
```diff
@@ -1259,7 +1338,7 @@ void py_module(py::module& module) {
         ttnn::expm1_bw,
         R"doc(Performs backward operations for exp2 on :attr:`input_tensor` with given :attr:`grad_tensor`)doc");
 
-    detail::bind_unary_backward_op(
+    detail::bind_unary_backward_op_overload(
         module,
```
Comment (eyonland) [inline ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward_pybind.hpp:1342]: Why are we renaming these bind_unary_backward_op ?

**Review by TT-BrianLiu — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/eltwise/complex/complex.hpp [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    private:
+        std::array<Tensor,2> m_real_imag;
+
+    public:
+
+        ComplexTensor(std::array<Tensor,2> val): m_real_imag(val) {
+            TT_ASSERT( m_real_imag[0].get_legacy_shape() == m_real_imag[1].get_legacy_shape() , "Tensor shapes of real and imag should be identical");
+        }
+
+        const Tensor& operator[](uint32_t index) const {
```
Comment (dmakoviichuk-tt) [inline ttnn/cpp/ttnn/operations/eltwise/complex/complex.hpp:23]: better to add check and throw exception if index is out of range.

**Review by dmakoviichuk-tt — APPROVED** [review]
One small comment from me. And approved.

File: ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward_pybind.hpp (hunk 5 above)
Comment (Aswinmcw) [reply ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward_pybind.hpp:1342]: `bind_unary_backward_op` contains normal operations that doesnot require overloading, and this `bind_unary_backward_op_overload` will have overloading struct that requires different pybind to function properly

**Review by eyonland — APPROVED** [review]

**Review by eyonland — APPROVED** [review]

**Pull request merged**